### ISSUE TITLE
Partial ordering of constrained member functions

### ISSUE THREAD

**Issue by PiliLatiesa** [llvm/llvm-project#58165]
```c++
template<typename>
constexpr bool b = true;

template<typename T>
concept C = b<T>;

template<typename T, typename U>
class foo
{
public:
  void bar() const requires C<T> {} // #1
  void bar() const requires (C<T> && C<U>) {} // Subsumes #1
};

template<typename T, typename U>
void bar(foo<T, U> const &) requires C<T> {} // #2

template<typename T, typename U>
void bar(foo<T, U> const &) requires (C<T> && C<U>) {} // Subsumes #2

int main()
{
  foo<int, int> x;
  x.bar(); // Clang reports an ambiguity
  bar(x);  // Clang does not complain
  return 0;
}
```

**Comment by shafik** [comment]
gobolt showing clang rejects and gcc accepts the code: https://godbolt.org/z/n1GsYKvGs

**Comment by PiliLatiesa** [comment]
If the member functions are templates:

```c++
template<typename>
constexpr bool b = true;

template<typename T>
concept C = b<T>;

template<typename T, typename U>
class foo
{
public:
  template<typename = void>
  void bar() const requires C<T> {} // #1

  template<typename = void>
  void bar() const requires (C<T> && C<U>) {} // Subsumes #1
};

template<typename T, typename U>
void bar(foo<T, U> const &) requires C<T> {} // #2

template<typename T, typename U>
void bar(foo<T, U> const &) requires (C<T> && C<U>) {} // Subsumes #2

int main()
{
  foo<int, int> x;
  x.bar(); // Clang reports an ambiguity
  bar(x);  // Clang does not complain
  return 0;
}
```
then Clang trunk accepts the testcase, but Clang 15 rejects it.

Perhaps @erichkeane 's patch on deferred evaluation of constraints made a difference here.

**Comment by erichkeane** [comment]
Maybe?  We HAVE had work done in the partial ordering of constrained functions since 15 though.

**Comment by erichkeane** [comment]
See https://reviews.llvm.org/D128750  Perhaps @yuanfang-chen  has some opinions on it?

**Comment by yuanfang-chen** [comment]
I think the test case should be accepted. 

When picking the overloaded non-template function, constraints partial ordering applied by 
https://eel.is/c++draft/over.match.best#general-2.6

**Comment by PiliLatiesa** [comment]
I believe this has been already fixed.

**Comment by mizvekov** [comment]
This has been fixed since clang 16.

https://godbolt.org/z/3vj5zP1ne
